src/UserSpaceInstrumentation/Trampoline.cpp

```diff
@@ -161,4 +176,143 @@ ErrorMessageOr<uint64_t> AllocateMemoryForTrampolines(pid_t pid, const AddressRa
+  constexpr uint64_t kAbsMinInt32AsUint64 =
+      static_cast<uint64_t>(-static_cast<int64_t>(std::numeric_limits<int32_t>::min()));
+  if ((a > b) && (a - b > kAbsMaxInt32AsUint64)) {
+    return ErrorMessage("Difference is larger than +2GB.");
+  }
+  if ((b > a) && (b - a > kAbsMinInt32AsUint64)) {
+    return ErrorMessage("Difference is larger than -2GB.");
+  }
+  return a - b;
+}
+
+ErrorMessageOr<RelocatedInstruction> RelocateInstruction(cs_insn* instruction, uint64_t old_address,
+                                                         uint64_t new_address) {
+  RelocatedInstruction result;
+  if ((instruction->detail->x86.modrm & 0xC7) == 0x05) {
+    // The encoding of an x86 instruction contains instruction prefixes, an opcode, the modrm and
+    // sib bytes, 1, 2 or 4 bytes address displacement and 1, 2 or 4 bytes of immediate data.
+    // Most of these are optional - at least one byte of opcode needs to be present.
+    // Many instructions that refer to an operand in memory have an addressing-form specifier byte
+    // (called the modrm byte) following the primary opcode.
+    // In case (modrm & 0xC7 == 0x05) this modrm bytes encodes a memory operand that is computed as
+    // the rip of the next instruction plus the 32 bit offset encoded in the four address
+    // displacement bytes of the instruction.
+    // See "Intel 64 and IA-32 Architectures Software Developer’s Manual Vol. 2A" Chapter 2.1.
+    // Specifically table 2-2.
+    const int32_t old_displacement = *absl::bit_cast<int32_t*>(
+        instruction->bytes + instruction->detail->x86.encoding.disp_offset);
+    const uint64_t old_absolute_address = old_address + instruction->size + old_displacement;
+    ErrorMessageOr<int32_t> new_displacement_or_error =
+        AddressDifferenceAsInt32(old_absolute_address, new_address + instruction->size);
+    if (new_displacement_or_error.has_error()) {
+      return ErrorMessage(absl::StrFormat(
+          "While trying to relocate an instruction with rip relative addressing the target was out "
+          "of range from the trampoline. old address: %#x, new address: %#x, instruction: %s",
+          old_address, new_address, InstructionBytesAsString(instruction)));
+    }
+    result.code.resize(instruction->size);
+    memcpy(result.code.data(), instruction->bytes, instruction->size);
+    *absl::bit_cast<int32_t*>(result.code.data() + instruction->detail->x86.encoding.disp_offset) =
+        new_displacement_or_error.value();
+  } else if (instruction->detail->x86.opcode[0] == 0xeb ||
+             instruction->detail->x86.opcode[0] == 0xe9) {
+    // This handles unconditional jump to relative immediate parameter (32 bit or 8 bit).
+    // Examples of original code
+    // (endless loop doing nothing)
+    // nop                          90
+    // jmp 0xfc                     eb fc
+    // (jump to a 32 bit offset)
+    // jmp 0x01020304               e9 01 20 03 04
```

Comment:
I would make it clearer these are two separate examples, I was somehow confused at the beginning.

Reply:
I just removed one of the examples. The difference between 8 and 32 bit offsets shpuld be small enough.